### PR TITLE
Python bindings with Swig

### PR DESCRIPTION
**Pull request by rdeits**
Not ready for merge yet, but I've cleaned up and updated my swig branch to use the current master of Drake. I've also brought in the work I did in https://github.com/rdeits/swig-autodiff to add autodiff handling to the Python bindings. I'm really happy with the result. To do an autodiff call to forwardKin, for example, you can do:

```
q = newAutoDiff(np.zeros((7,1)), np.eye(7, 7))
v = newAutoDiff(np.zeros((7,1)), np.zeros((7, 7)))
kinsol = r.doKinematics(q, v)

point = np.ones((3,1))
p = r.forwardKin(kinsol, point, 2, 0, 0)
print(p.value())
print(p.derivatives())
```

which gives:

```
[[ 1.]
 [ 1.]
 [ 1.]]
[[ 1.  0.  0.  0.  1. -1.  1.]
 [ 0.  1.  0. -1.  0.  1.  0.]
 [ 0.  0.  1.  1. -1.  0. -1.]]
```

Todo: 

- [ ] adapt this to the new Drake System model
- [x] make the python examples unit tests in CTest
- [ ] test python3 support. I know that the components in https://github.com/rdeits/swig-autodiff and https://github.com/rdeits/swig-eigen-numpy work in python3, but I haven't tried all of Drake yet
- [ ] add more functions to the python wrappers
- [ ] consider adding support for fixed max size AutoDiff types. I've actually already done that in https://github.com/rdeits/swig-autodiff but haven't yet turned it on for Drake, since it means even more instantiations of the various functions

### CODE REVIEW

**Comment by rdeits** [comment]
Another thought: in order to use swig to generate Matlab bindings, we have to build our own version of swig. That's not a problem on osx or Linux, but it won't work on Windows (the swig build requires cygwin or msys). If we don't use swig for our Matlab bindings, then that's no problem. Otherwise, I guess we'd have to distribute a precompiled version of swig with Matlab for Windows.

**Comment by rdeits** [comment]
@RussTedrake drake005 is failing to build because it needs `autoconf`, `automake` and `libtool` from homebrew, due to swig's build system (which, again, we only *need* if we're generating swig matlab bindings). I assume the windows build servers will all fail as well. 

So, we have a few options:

1. switch to the standard systemwide swig installation
  * this should work fine on Windows, and it simplifies drake's build, but it won't support Matlab at all
1. use standard swig in Windows and swig+matlab elsewhere
  * not terribly attractive if we *actually* care about windows being a useful platform
1. require a unix-y environment on Windows and use swig's automake-based build everywhere
  * this would give us Python & Matlab bindings everywhere, but at the cost of requiring an entire cygwin or msys environment for everyone
1. create a cmake build for swig on Windows
  * this isn't actually too crazy. There was an effort between SWIG and Kitware in 2003 to port SWIG to cmake, but it stalled: http://thread.gmane.org/gmane.comp.programming.swig.devel/21606/focus=21611 There is still a branch floating around, though: https://github.com/swig/swig/blob/cmake/CMakeLists.txt

I think that, for the sake of our sanity, I'll just switch to using the system-provided version of swig and only create python bindings for this pull request (option 1). That will fix the blocking issues. If the python bindings end up being useful, it should be easy to switch on the matlab bindings on OSX and Linux, and perhaps we can get some help in dealing with the windows issues. 

Somehow, I've managed to create a problem for myself that combines Windows, automake, *and* Matlab all at the same time! :wink:

**Comment by rdeits** [comment]
Okay, the swig wrappers work, and my first two python unit tests are passing on all four Travis builds. Woohoo! 

drake005 is failing because it's still trying to build swig, and still missing autoconf, etc. but that should be easy to fix.

File: drake/bindings/swig/rbm.i [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  AutoDiffWrapper<Eigen::VectorXd, Eigen::Dynamic, 1> forwardKin(
+      const KinematicsCache<Eigen::AutoDiffScalar<Eigen::VectorXd> > &cache, const Eigen::Matrix<double, SPACE_DIMENSION, 1> &points, int current_body_or_frame_ind, int new_body_or_frame_ind, int rotation_type) const
+  {
+    return $self->forwardKin(cache, points, current_body_or_frame_ind, new_body_or_frame_ind, rotation_type);
+  }
+
+  Eigen::Matrix<double, SPACE_DIMENSION, 1> centerOfMass(KinematicsCache<double> &cache, const std::set<int> &robotnum = default_robot_num_set) const {
+    return $self->centerOfMass(cache, robotnum);
+  }
```
Comment (rdeits) [inline drake/bindings/swig/rbm.i]: Every templated method from RigidBodyTree that we want to be accessible from Python will only require a few new lines in this file.

File: drake/systems/plants/IKoptions.h [hunk 2]
```diff
@@ -63,7 +63,7 @@ class drakeIKoptions_DLLEXPORT IKoptions
     void setqd0(const Eigen::VectorXd &lb, const Eigen::VectorXd &ub);
     void setqdf(const Eigen::VectorXd &lb, const Eigen::VectorXd &ub);
     void setAdditionaltSamples(const Eigen::RowVectorXd &t_samples);
-    void updateRobot(const RigidBodyTree * robot);
```
Comment (rdeits) [inline drake/systems/plants/IKoptions.h]: the `const` version of this method wasn't implemented in the .cpp file, and the duplication in the header was causing weird swig issues

**Comment by RussTedrake** [comment]
just installed the required deps on drake005.

@rdeits , @tkoolen  and others -- what do you think about the following proposal for pushing through swig + matlab:  I think we should focus first on swigifying the RigidBodyTree class.  We have RigidBodyManipulator.m in matlab.  I think we could try to autogenerate wrappers for a brand-new matlab class @RigidBodyTree , which is a handle class and is not a DrakeSystem, who's sole purpose is to wrap the C++ interface.  If we can really autogen all the bindings with swig, then we could even potentially write @RigidBodyTree into the pod-build/matlab directory, instead of in source.   Then we could start slowing moving over the examples which use the tree logic (and do not require the DrakeSystem part) as a way of transitioning to the new interface.  I think that the RigidBodyTree class will be easier to swig than the systems classes, and will be immediately useful.

**Comment by RussTedrake** [comment]
@rdeits -- would love to watch your progress on this, but for my sanity maybe we should not leave it as an open pull request?  please re-open when appropriate.

**Comment by rdeits** [comment]
@RussTedrake agreed: RigidBodyTree seems like the right first target. I've rebased my work back onto the newest master, and I've got Python and Matlab bindings working. Due to the obnoxious LCM packaging issues, I can't call the bindings package "drake", and "pydrake" is not a good name for the Matlab bindings. So, until we have a better idea, the bindings package is `dragon` :wink: 

 The interfaces are almost exactly the same:

python: 

```python
r = dragon.rbtree.RigidBodyTree(os.path.join(dragon.getDrakePath(), "examples/Pendulum/Pendulum.urdf"))
print("num positions:", r.num_positions)
print("num velocities:", r.num_velocities)

kinsol = r.doKinematics(np.zeros((7,1)), np.zeros((7,1)))

p = r.forwardKin(kinsol, np.zeros((3,1)), 0, 1, 0)
print(p)
```

matlab: 

```matlab
r = dragon.RigidBodyTree('../../drake/examples/Pendulum/Pendulum.urdf')

Warning: joint dynamics xml tag is parsed, but not included in the dynamics methods yet.

r =

<Swig object, ptr=140612128157280>
>> r.num_positions

ans =

                    7

>> r.num_velocities

ans =

                    7

>> kinsol = r.doKinematics(zeros(7,1), zeros(7,1))

kinsol =

<Swig object, ptr=140612148204272>
>> p = r.forwardKin(kinsol, zeros(3,1), 0, 1, 0)

p =

     0
     0
     0
```

**Pull request closed without merging**